portalnet/src/overlay_service.rs

```diff
@@ -758,58 +758,147 @@ where
+                let (callback, content_key) = match query_info.query_type {
+                    QueryType::FindContent { callback, target } => (callback, target),
+                    _ => {
+                        error!("Query type mismatch");
+                        return;
+                    }
                 };
 
-                if let QueryType::FindContent {
-                    callback: Some(callback),
-                    target: content_key,
-                } = query_info.query_type
-                {
-                    let response = (content.clone(), query_info.trace);
-                    // Send (possibly `None`) content on callback channel.
-                    if let Err(err) = callback.send(response) {
-                        error!(
-                            query.id = %query_id,
-                            error = ?err,
-                            "Error sending FindContent query result to callback",
-                        );
+                match query.into_result() {
+                    FindContentQueryResult::ClosestNodes(_closest_nodes) => {
+                        if let Some(responder) = callback {
+                            let _ = responder.send((None, query_info.trace));
+                        }
                     }
+                    FindContentQueryResult::Content {
+                        content,
+                        nodes_to_poke,
+                    } => {
+                        let validator = self.validator.clone();
+                        let store = self.store.clone();
+                        let kbuckets = self.kbuckets.clone();
+                        let command_tx = self.command_tx.clone();
+                        tokio::spawn(async move {
+                            Self::process_received_content(
+                                kbuckets,
+                                command_tx,
+                                validator,
+                                store,
+                                (content.clone(), true),
+                                content_key,
+                                callback,
+                                query_info.trace,
+                                nodes_to_poke,
+                            );
+                        });
+                    }
+                    FindContentQueryResult::Utp {
+                        connection_id,
+                        peer,
+                        nodes_to_poke,
+                    } => {
+                        let metrics = self.metrics.clone();
+                        let utp = self.utp_socket.clone();
+                        let id = u16::from_be(connection_id);
+                        let key = kbucket::Key::from(peer);
+                        let source = match self.kbuckets.write().entry(&key) {
+                            kbucket::Entry::Present(entry, _) => entry.value().enr(),
+                            _ => {
```

Comment:
```suggestion
                        let source = match self.find_enr(&peer) {
                            Some(enr) => enr,
                            None => {
```

I was getting a lot of `Received uTP payload from unknown peer` warnings before making this change

Reply:
Cool, nice catch. `find_enr` seems like the better mechanism here. We also do `kbuckets.write()....` inside `poke_content` so it seems like it might be worth updating that also to use `find_enr`?

Reply:
@mrferris Ahh, nvm I was wrong, inside `poke_content` we don't have access to `self` so `find_enr()` is inaccessible